yt_dlp/extractor/adultswim.py

```diff
@@ -196,3 +205,154 @@ def _real_extract(self, url):
+        'info_dict': {
+            'id': 'rick-and-morty',
+            'title': 'Rick and Morty',
+            'description': 'An infinite loop of Rick and Morty. You\'re welcome. (Marathon available in select regions)',
+        },
+        'params': {
+            # m3u8 download
+            'skip_download': True,
+        },
+        'playlist_mincount': 40,
+    }]
+
+    def _live_hls_fragments(self, episode_start_time, episode_duration, video_id, hls_url, content):
+        FRAGMENT_DURATION = 10.010
+
+        sleep_until = episode_start_time + min(60, episode_duration)
+        if time.time() - sleep_until < 0:
+            format_dur = lambda dur: '%02d:%02d:%02d' % timetuple_from_msec(dur * 1000)[:-1]
+            last_msg = ''
+
+            def progress(msg):
+                nonlocal last_msg
+                self.to_screen(msg + ' ' * (len(last_msg) - len(msg)) + '\r', skip_eol=True)
+                last_msg = msg
+
+            while time.time() <= sleep_until:
+                progress('Waiting for next episode to air (%s) - Press Ctrl+C to cancel' % format_dur(sleep_until - time.time()))
+                time.sleep(1)
+            progress('')
+
+            content = self._download_webpage(hls_url, video_id, note='Downloading m3u8 manifest')
+        elif time.time() - sleep_until > episode_duration:
+            self.report_warning('Skipping episode as new episode has already aired')
+            return []
+
+        fragments, error_msg = HlsFD._parse_m3u8(content, {'url': hls_url})
+        if not fragments:
+            self.report_warning(error_msg)
+            return []
+
+        for f in reversed(fragments):
+            match = re.search(r'^https?:\/\/adultswim-vodlive.cdn.turner.com\/.*\/seg[^_]+_(?P<index>\d+)\.ts$', f['url'])
+            if match:
+                fragment_template = f
+                break
+        else:
+            self.report_warning('Could not find any valid stream segments')
+            return []
+
+        digit_str_index, digit_str_length = match.span('index')[0], len(match.group('index'))
+        fragment_url_template = fragment_template['url'][:digit_str_index] + '%s' + fragment_template['url'][digit_str_index + digit_str_length:]
+        fragment_count = math.ceil(episode_duration / FRAGMENT_DURATION)
+        for i in reversed(range(fragment_count)):
+            try:
+                self._downloader.urlopen(fragment_url_template % '{0:0{width}}'.format(i, width=digit_str_length))
+                break
+            except Exception:
+                fragment_count -= 1
+
+        return [{'frag_index': i,
+                 'url': fragment_url_template % '{0:0{width}}'.format(i, width=digit_str_length),
+                 'decrypt_info': fragment_template['decrypt_info'],
+                 'byte_range': fragment_template['byte_range'],
+                 'media_sequence': fragment_template['media_sequence']} for i in range(fragment_count)]
```

Comment:
that would make quite a lot of headers requests, wouldn't it be better to limit those to only the last segments where it is unsure whether the file is available?

Reply:
i feel like it looks cleaner when there are two for loops, otherwise it would look like this:
```
        available = False
        for i in reversed(range(fragment_count)):
            fragment_url = fragment_url_template % f'{i:0{digit_str_length}}'
            if not available and not self._request_webpage(HEADRequest(fragment_url_template % f'{i:0{digit_str_length}}'),
                video_id, note=f'Determining availability of segment (Segment Length: {fragment_count})', errnote=False):
                continue
            available = True

            yield {'frag_index': i,
                   'url': fragment_url,
                   'decrypt_info': fragment_template['decrypt_info'],
                   'byte_range': fragment_template['byte_range'],
                   'media_sequence': fragment_template['media_sequence']}
```

Reply:
hm.. I see your point. In that case, 2 separate loops makes sense. But the other changes (fstrings, request_webpage) are still applicable

Reply:
> it will look like this then:
> 
> ```python
>         available = False
>         for i in reversed(range(fragment_count)):
>             fragment_url = fragment_url_template % f'{i:0{digit_str_length}}'
>             if not available and not self._request_webpage(HEADRequest(fragment_url_template % f'{i:0{digit_str_length}}'),
>                 video_id, note=f'Determining availability of segments (Segment Length: {fragment_count})', errnote=False):
>                 continue
>             available = True
> 
>             yield {'frag_index': i,
>                    'url': fragment_url,
>                    'decrypt_info': fragment_template['decrypt_info'],
>                    'byte_range': fragment_template['byte_range'],
>                    'media_sequence': fragment_template['media_sequence']}
> ```

This will yield the fragments in reverse, which is not what we want